File: leetcode/product-of-array-except-self.java
//https://leetcode.com/problems/product-of-array-except-self/

class Solution {
    public int[] productExceptSelf(int[] nums) {
        int[] sums = new int[nums.length];
        
        //find products going 0 through n-1
        int currProd = 1;
        for(int i = 0; i<nums.length; i++){
            sums[i] = currProd;
            currProd*=nums[i];
        }
        
        //include products from n-1 through 0
        currProd = 1;
        for(int i = nums.length-1; i>=0; i--){
            sums[i] *= currProd;
            currProd*=nums[i];
        }
        
        return sums;
    }
}

File: leetcode/product_of_array_except_self.py
# https://leetcode.com/problems/product-of-array-except-self/description/


class Solution:
    def productExceptSelf(self, nums: list[int]) -> list[int]:
        arr: list[int] = [1 for i in range(len(nums))]

        for i in range(1, len(nums)):
            arr[i] = nums[i - 1] * arr[i - 1]

        prod: int = 1
        for i in range(len(nums) - 2, -1, -1):
            prod *= nums[i + 1]
            arr[i] *= prod

        return arr
